src/vsphere_cpi/spec/integration/gpu_parallel_spec.rb

```diff
@@ -0,0 +1,310 @@
+require 'integration/spec_helper'
+require 'rspec/expectations'
+
+GRAPHICS_VENDOR_ID = 4318.freeze
+
+RSpec::Matchers.define :have_number_of_GPU_eql_to do |num_gpu|
+  match do |actual|
+    num_gpu  == actual.count do |device|
+      !device.backing.nil? && device.backing.is_a?(VimSdk::Vim::Vm::Device::VirtualPCIPassthrough::DeviceBackingInfo) && device.backing.vendor_id == GRAPHICS_VENDOR_ID
+    end
+  end
+end
+
+#INCLUDES TESTS for parallel vm calls and tests for gpu on hosts in different clusters
+
+#have a before before a context
+#for second context one host in one cluster then the other one in other cluster
+describe 'cloud_properties related to creation of GPU attached VMs' do
+  before (:all) do
+    @datacenter_name = fetch_and_verify_datacenter('BOSH_VSPHERE_CPI_DATACENTER')
+    @cluster_name = fetch_and_verify_cluster('BOSH_VSPHERE_CPI_CLUSTER')
+    @second_cluster_name = fetch_and_verify_cluster('BOSH_VSPHERE_CPI_SECOND_CLUSTER')
+    @gpu_host_datastore_one = fetch_property('BOSH_VSPHERE_CPI_SHARED_DATASTORE')
+    @host_1 = '10.114.22.249' #4 GRID K1 GPUs
+    @host_2 = '10.114.22.29' #2 GRID K2 GPUs
+  end
+
+  let(:base_vm_type) do
+    {
+        'ram' => 512,
+        'disk' => 2048,
+        'cpu' => 1,
+    }
+  end
+
+  let(:options) do
+    options = cpi_options(
+        'datacenters' => [
+            {
+                'name' => @datacenter_name,
+                'clusters' => [
+                    {
+                        @cluster_name => {},
+                    }
+                ]
+            }
+        ]
+    )
+  end
+
+  let(:cpi) do
+    VSphereCloud::Cloud.new(options)
+  end
+
+  # Both hosts being on the second cluster tests
+  context 'when vm_type specifies a single cluster that has one of the hosts with GPUs' do
+    let(:vm_type_single_cluster) do
+      base_vm_type.merge(
+          'datacenters' => [
+              {
+                  'name' => @datacenter_name,
+                  'clusters' => [
+                      {
+                          @second_cluster_name => {}
+                      },
+                  ]
+              }
+          ]
+      )
+    end
+    context 'when ephemeral datastore is accessible from GPU host' do
+      let(:vm_type_single_cluster_acc_ds) do
+        vm_type_single_cluster.merge(
+            'datastores' => [@gpu_host_datastore_one]
+        )
+      end
+
+      # parallel vm calls
+      context 'when vm_type has two clusters specifies 1 GPU and we create 6 vms in parallel' do
```

Comment:
Looks like context are not arranged properly, the outermost context says ' when vm_type specifies single cluster and this says when it has 2 clusters'. This context should not be nested inside the outermost context

Reply:
going through all the context and let is confusing, this needs to be cleaned up